### PR TITLE
Validate PING and STATE connections at join time

### PR DESCRIPTION
**Pull request by DaveCTurner**
When a node requests to join the cluster we perform some basic
validation to ensure that it makes sense to process its join request.
Today we verify its version and that it can read a recent cluster state
without errors. This commit adds some extra validation that the
connections needed for the node to join the cluster are not broken (i.e.
silently dropping packets). We notice broken channels eventually but it
might take many minutes, and until we notice the node may be joining and
leaving the cluster repeatedly which is quite disruptive. In more
detail:

- If the established `STATE` channel from the master to the joining node
is broken then today we let the node join and then remove it a couple of
minutes later for lagging.

- If the established `PING` channel from the master to the joining node
is broken then today we let the node join and then remove it 30s later
for ping failures.

- If the established `PING` channel from the joining node to the master
is broken then today the node will join but keep on rejoining
unnecessarily, believing the master to be failing to respond to its
pings.

With this commit we ensure that the joining node has a good `PING`
channel to the master, and that the master has good `STATE` and `PING`
channels back to the joining node, before starting to process its join.
We do this by sending the join request on the `PING` channel, then
sending the validation state back on the `STATE` channel and a new
lightweight message on the `PING` channel in parallel.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Failure is https://github.com/elastic/elasticsearch/issues/77751; @elasticmachine please run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 1]
```diff
@@ -511,42 +500,118 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
...
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(Empty empty) {
+                            validateListener.onFailure(e);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e2) {
+                            e.addSuppressed(e2);
+                            validateListener.onFailure(e);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:547]: nit: I think I would prefer to swap these, seems like the join-validate exception failure is "more important" than the ping one.
```suggestion
                            e2.addSuppressed(e);
                            validateListener.onFailure(e2);
```

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 2]
```diff
@@ -511,42 +500,118 @@ private void handleJoinRequest(JoinRequest joinRequest, ActionListener<Void> joi
...
+            listener.onResponse(Empty.INSTANCE);
+        }
+    }
+
     private void processJoinRequest(JoinRequest joinRequest, ActionListener<Void> joinListener) {
         final Optional<Join> optionalJoin = joinRequest.getOptionalJoin();
-        synchronized (mutex) {
-            updateMaxTermSeen(joinRequest.getTerm());
+        try {
+            synchronized (mutex) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:600]: It is clearly fine now, but with the added complexity, I wonder if we should assert that we are not on the transport thread here?

**Review by henningandersen — APPROVED** [review]
LGTM, thanks for addressing this.

**Pull request merged**